### ISSUE TITLE
BenchmarkDotNet doesn't seem to install all necessary dependencies

### ISSUE THREAD

**Issue by smoothdeveloper** [fsprojects/Paket#2109]
### Description

paket.dependencies:

```
source https://www.nuget.org/api/v2
strategy: max
redirects: on
framework: >= net40

nuget BenchmarkDotNet = 0.10.1
```

in the nuspec of BenchmarkDotNet.Core (which is transitive dependency and installed), it seems to mention `System.Threading.Tasks.Extensions`
```xml
<?xml version="1.0" encoding="utf-8"?>
<package xmlns="http://schemas.microsoft.com/packaging/2012/06/nuspec.xsd">
    <metadata>
        <id>BenchmarkDotNet.Core</id>
        <version>0.10.1</version>
        <title>BenchmarkDotNet.Core</title>
        <authors>.NET Foundation and contributors</authors>
        <owners>.NET Foundation and contributors</owners>
        <licenseUrl>https://github.com/dotnet/BenchmarkDotNet/blob/master/LICENSE.md</licenseUrl>
        <projectUrl>https://github.com/dotnet/BenchmarkDotNet</projectUrl>
        <iconUrl>https://raw.githubusercontent.com/dotnet/BenchmarkDotNet/master/docs/guide/logo.png</iconUrl>
        <requireLicenseAcceptance>false</requireLicenseAcceptance>
        <description>Powerful .NET library for benchmarking</description>
        <copyright>.NET Foundation and contributors</copyright>
        <language>en-US</language>
        <tags>benchmark benchmarking performance</tags>
        <dependencies>
            <group targetFramework=".NETCoreApp1.1">
                <dependency id="Microsoft.NETCore.App" version="1.1.0" />
                <dependency id="System.Runtime.Serialization.Primitives" version="4.1.1" />
                <dependency id="Microsoft.DotNet.InternalAbstractions" version="1.0.0" />
                <dependency id="System.Xml.XPath.XmlDocument" version="4.0.1" />
            </group>
            <group targetFramework=".NETFramework4.5">
                <dependency id="System.Threading.Tasks.Extensions" version="4.0.0" />
            </group>
        </dependencies>
        <frameworkAssemblies>
            <frameworkAssembly assemblyName="System.Management" targetFramework=".NETFramework4.5" />
            <frameworkAssembly assemblyName="System.Xml" targetFramework=".NETFramework4.5" />
        </frameworkAssemblies>
    </metadata>
</package>
```

on install, csproj targetting .net 4.5 doesn't have the `System.Threading.Tasks.Extensions` assembly reference added which I think is a bug.

Minimal project for addition in integration tests: [paketbenchmarkdotnet.zip](https://github.com/fsprojects/Paket/files/711872/paketbenchmarkdotnet.zip)

**Comment by vbfox** [comment]
Two Problems here :

1. Paket doesn't yet know anything about "netcoreapp1.1"
2. When Paket see an unknown target framework for a dependency it assume that it's for ".net 1.0 +"

I'll try to fix the first one but the second one might need input from @forki : 
* What to do here ? stop paket completly and require the user to fill an issue ? 
* If we just warn we need to do something about the `.json` cache : A new version of paket might generate a different cache and keeping old cached files is sure to confuse users. (Non working packages that would work if the cache was cleared)

----

`NugetV2.fs`:

```fsharp
        let split (d : string) =
            let a = d.Split ':'
            PackageName a.[0],
            VersionRequirement.Parse(if a.Length > 1 then a.[1] else "0"),
            (if a.Length > 2 && a.[2] <> "" then
                 if String.startsWithIgnoreCase "portable" a.[2] then
                    [ yield FrameworkRestriction.Portable a.[2]]
                 else
                     match FrameworkDetection.Extract a.[2] with
                     | Some x -> [ FrameworkRestriction.Exactly x ]
                     | None -> [] // XXX: This line shouldn't just ignore that we weren't able to parse something !
             else [])
```

**Comment by vbfox** [comment]
BTW reflecting on that I don't think anything smart would scale into later versions for the cache.

The easiest way to go around it is to not include such references in the packages (So we ignore them) and use a version number in the json cache file name.

Such packages in the future will "just work" if nobody care about said unknown framework, and once we add them we'll bump the cache version number and they'll "just work" because the cache will be up to date.

It incur a cache re-generation each time we bump the value in paket but it just require internet acces and updating paket also require an internet access so it's not a big problem.

**Comment by smoothdeveloper** [comment]
@vbox thanks for feedback, in my case I'm using framework > net40 (only .net framework on windows), I'm thinking it should already work for that section

```xml
<group targetFramework=".NETFramework4.5">
    <dependency id="System.Threading.Tasks.Extensions" version="4.0.0" />
</group>
```

Is it failing because it doesn't understand the previous sibbling section?

**Comment by vbfox** [comment]
> Is it failing because it doesn't understand the previous sibbling section?

Yes essentially any section with a framework reference unknown to paket can cause that problem. **3.35.0** should solve this specific case but i'll look into solving it for good (So that unknown frameworks don't risk affecting users that don't use them)

**Comment by smoothdeveloper** [comment]
@vbfox understood, thanks for the clarification.

Yes if paket can handle those unknown sections (maybe log a warning in verbose mode) by just skipping those, I think that would solve the issue.

> stop paket completly and require the user to fill an issue ?

stopping paket completely is not going to help user, in my case I could work around by adding the reference myself, if paket stops, I'm going to be upset because I need paket to complete other tasks instead of stopping there.

I don't understand the details about the cache though, but my thinking is that the full .nupkg is there and under packages it is fully extracted, no matter which framework restriction or settings I have.

**Comment by vbfox** [comment]
> stopping paket completely is not going to help user, in my case I could work around by adding the reference myself, if paket stops, I'm going to be upset because I need paket to complete other tasks instead of stopping there.

I think we'll avoid that 😉 

> I don't understand the details about the cache though, but my thinking is that the full .nupkg is there and under packages it is fully extracted, no matter which framework restriction or settings I have.

Essentially paket query the nuget server (via OData or whatever nuget v3 protocol uses) and cache the result to avoid querying it again (It might not need to download the package at all in the end so can't depend on the nupkg).

But that cache is essentially a Json serialization of paket data-structures not the raw received data so if it was download with Paket vX that didn't know about "netstandard 42.5" it can't reference it. Once a version of paket is available with "netstandard 42.5" the cache need to be invalidated.

It should be noted that this cache problem show in **3.35.0** : If you update today and start using it on the same machine it failed before **BenchmarkDotNet** won't work because the cache isn't invalidated (there is no logic to do that right now)  and paket will still think that the .Core dll is needed on framework 1.0+. (After a clear cache it should work)
